### PR TITLE
1023: Dependabot Mass Merge

### LINKED ISSUE #1023: Dependabot Mass Merge / Improvements
**Description:**
Since Dependabot was implemented we have been churning through a significant backlog of package updates that were previously missed. Per our [documented git flow](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/Architecture%20Decision%20Record/009-git-workflow.md) we have been reviewing these PRs one at a time and putting them through the regular QASP review process. However, in the case of large volumes of these PRs this is very time consuming and a lot of these updates are minor version increases which don't change any of our functionality.

We have discussed a process where these dependency PRs could be bulk merged into a single branch and reviewed as a whole whenever they exceed a certain volume. Once all dependency updates were in one branch the typical review process would be followed. After the backlog has been cleared these dependency updates should be a much lower frequency so they could be handled in the typical git flow as needed.

Additionally, Dependabot PRs currently are created with no reviewers - we should update the configuration so PRs are opened with reviewers from the Raft team.

Another issue seen with Dependabot PRs on the Python side specifically, is that many of our dependencies don't have a strict equality setting for the version number - leading to unintended upgrades of unrelated packages when generating a lock file for a distinct package update. After we have merged all of these updates in we should set all the versions to strict release numbers and allow Dependabot to handle upgrading them individually.

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [x] Dependabot update backlog is cleared out - ie all dependencies are up to date
- [x] Dependabot ADR is updated to cover this option in case it needs to be done again in the future
- [x] The app has been manually tested and confirmed to work after upgrading all dependencies queued by Dependabot
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [x] Open a PR to update the Dependabot ADR to document this process
- [x] Create a temporary branch to hold all dependabot updates - branched from raft-tdp-main
- [x] Close/Merge any open Dependabot updates
- [x] Open a PR to update the Dependabot config to target the branch from the first task and increase the open PR limit to 50
- [x] Update the Dependabot config to set reviewers to @raft-tech/tdrs-developers
- [x] Merge all Dependabot PRs into the branch from the first task
- [x] Revert the Dependabot config prior to merging
- [x] Open a new PR to raft-tdp-main with all dependency updates
- [x] Update versions in the Pipfile to use strictly equals (==) 
- [x] Thoroughly test the application before submitting to QASP
- [x] Run Testing Checklist and confirm all tests pass

**Notes:**

**Supporting Documentation:**

- [Dependabot target branch config](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/configuration-options-for-dependency-updates#target-branch)
- [Dependabot open PR limit config](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/configuration-options-for-dependency-updates#open-pull-requests-limit)
- [Dependabot reviewers config](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/configuration-options-for-dependency-updates#reviewers)

**Open Questions:**

### PR DESCRIPTION
**Pull request by jtwillis92**
### Summary of Changes
Merges all dependabot updates into a single branch and updates any code necessary to ensure that the application works correctly on the latest versions of all dependencies.

Closes #1023 

### How to Test
* Pull this branch down locally and rebuild all containers
* Log in to the UI and navigate around the app to ensure everything works as expected
* Deploy this PR by assigning the `Deploy with CircleCI-{your target env}` label
* Navigate to the deployed site and ensure everything works as expected

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  - [x] Dependabot update backlog is cleared out - ie all dependencies are up to date
  - [x] The app has been manually tested and confirmed to work after upgrading all dependencies queued by Dependabot

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)
  + [x] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

N/A no changes to UX

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces backend code, is that code documented both inline and overall?
+ [x] If this PR introduces frontend code, is that code documented both inline and overall?
+ [x] If this PR introduces dependencies, are their licenses documented?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -331,6 +331,11 @@ jobs:
             elif [ "$CIRCLE_BRANCH" == "raft-tdp-main" ] ; then
               cd tdrs-frontend; bash <(curl -s https://codecov.io/bash)  -f ./coverage/lcov.info -F dev-frontend
             fi
+      - run:
+          name: "Install lib bindings for Cypress"
```
Comment (jtwillis92) [inline .circleci/config.yml:335]: This additional package was needed to get Cypress to work on version 7.7: [ref](https://docs.cypress.io/guides/continuous-integration/introduction#Dependencies)

File: .github/dependabot.yml [hunk 2]
```diff
@@ -7,8 +7,7 @@ updates:
     directory: "/tdrs-frontend"
     schedule:
       interval: "daily"
-    target-branch: "dependabot-mass-merge"
-    open-pull-requests-limit: 50
+    target-branch: "raft-tdp-main"
```
Comment (jtwillis92) [inline .github/dependabot.yml:10]: Sets dependabot config back to normal so we can process future updates individually.

File: tdrs-backend/Pipfile [hunk 3]
```diff
@@ -4,47 +4,46 @@ url = "https://pypi.org/simple"
...
-django-unique-upload = "~=0.2.1"
-djangorestframework = "~=3.12.2"
-dj-database-url = "~=0.5.0"
-gunicorn = "~=20.0.4"
-ipdb = "~=0.13.9"
-ipython = "~=7.24.1"
-jwcrypto = "~=0.8"
-markdown = "~=3.3.3"
-psycopg2-binary = "~=2.8.6"
+boto3 = "==1.17.112"
```
Comment (jtwillis92) [inline tdrs-backend/Pipfile:23]: I alphabetized this for OCD reasons but the notable change here is that the version numbers were all updated to match what is currently in the Pipfile.lock and set to use a strict equality for the version number. This ensures that future dependabot PRs for Python dependencies won't update unrelated packages when building a new lock file.

File: tdrs-frontend/package.json [hunk 4]
```diff
@@ -9,23 +9,23 @@
     "axios": "^0.21.1",
     "classnames": "^2.3.0",
     "connected-react-router": "^6.8.0",
-    "file-type": "^16.3.0",
+    "file-type": "^16.5.1",
```
Comment (jtwillis92) [inline tdrs-frontend/package.json:12]: Yarn doesn't seem to have the same issue as pipenv for updating unrelated packages so these `^` versions are fine here.

File: tdrs-frontend/src/actions/reports.js [hunk 5]
```diff
@@ -16,112 +16,118 @@ export const SET_FILE_LIST = 'SET_FILE_LIST'
 export const FETCH_FILE_LIST_ERROR = 'FETCH_FILE_LIST_ERROR'
 export const DOWNLOAD_DIALOG_OPEN = 'DOWNLOAD_DIALOG_OPEN'
 
-export const clearFile = ({ section }) => (dispatch) => {
-  dispatch({ type: CLEAR_FILE, payload: { section } })
-}
+export const clearFile =
```
Comment (jtwillis92) [inline tdrs-frontend/src/actions/reports.js:19]: The prettier update caused some linter errors on this file so I used `eslint src/ --fix` to get this file to pass the linter.

File: tdrs-frontend/src/actions/requestAccess.js [hunk 6]
```diff
@@ -7,28 +7,30 @@ export const SET_REQUEST_ACCESS = 'SET_REQUEST_ACCESS'
...
-    const { data } = await (await axiosInstance).patch(URL, user, {
-      withCredentials: true,
-    })
-
-    if (data) {
-      dispatch({ type: SET_REQUEST_ACCESS })
-      dispatch({
-        type: SET_AUTH,
-        payload: { user: data },
+export const requestAccess =
```
Comment (jtwillis92) [inline tdrs-frontend/src/actions/requestAccess.js:10]: The prettier update caused some linter errors on this file so I used `eslint src/ --fix` to get this file to pass the linter.

File: tdrs-backend/Pipfile (hunk 3 above)
Comment (abottoms-coder) [reply tdrs-backend/Pipfile:23]: Thank you for this! Driving me up a wall

File: tdrs-frontend/src/actions/reports.js (hunk 5 above)
Comment (abottoms-coder) [reply tdrs-frontend/src/actions/reports.js:19]: Maybe it's just how github is rendering it or my unfamiliarity with JS but does the indentation seem off?

**Review by abottoms-coder — APPROVED** [review]

File: tdrs-frontend/src/actions/reports.js (hunk 5 above)
Comment (jtwillis92) [reply tdrs-frontend/src/actions/reports.js:19]: I think its github, it looks right in my IDE and the formatting matches what https://prettier.io/playground gives as well.

File: .circleci/config.yml (hunk 1 above)
Comment (jtwillis92) [reply .circleci/config.yml:335]: Note: We are on v8 now- there was a new version of Cypress available.

**Comment by jtwillis92** [comment]
Note: Following this PR we can expect 1-2 PRs a day from Dependabot. Specifically awscli and boto3 have very active release cadences, with a new version almost nightly. However, `awscli` is a dev dependency only used for testing so those can be merged as long as the tests pass without needing additional manual testing. Likewise, `boto3` includes a lot of functionality that we don't use so if no changes to the `s3` module are listed in the release notes that package also needs no further manual testing.

**Comment by ADPennington** [comment]
@jtwillis92 did you come across any dependencies that are no longer needed as part of this update process? Mostly just curious, since, eventually as part of the prod readiness work, i'd like to revisit #838

**Comment by jtwillis92** [comment]
> @jtwillis92 did you come across any dependencies that are no longer needed as part of this update process? Mostly just curious, since, eventually as part of the prod readiness work, i'd like to revisit #838

I did, `python-dotenv` was unused and removed in this PR. [ref](https://github.com/raft-tech/TANF-app/pull/1096/files#diff-c26ac7efe4156de068ca08f298f78465d9250f0bb63134f3e16e897e69266f45L42)

**Review by ADPennington — APPROVED** [review]
LGTM thanks @jtwillis92 👍🏾 

@abottoms-coder this is ready to merge. 


### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  + [x] Dependabot update backlog is cleared out - ie all dependencies are up to date
  + [x] The app has been manually tested and confirmed to work after upgrading all dependencies queued by Dependabot

also:
- [x] using strict equality setting for python dependencies to improve dependency mgmt
- [x] removed unused dependencies as part of the mass merge process
- [x] dependabot config updated to resume normal process. 

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)
  + [x] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

n/a

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*


> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces backend code, is that code documented both inline and overall?
+ [x] If this PR introduces frontend code, is that code documented both inline and overall?
+ [x] If this PR introduces dependencies, are their licenses documented?
https://github.com/raft-tech/TANF-app/pull/1096/files#r670718191
### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?
_no new issues detected_

**Pull request merged**